src/backends/policy_engine/souffle/datalog_lowering_visitor_test.cc

```diff
@@ -0,0 +1,243 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#include "datalog_lowering_visitor.h"
+
+#include "absl/strings/string_view.h"
+#include "src/common/testing/gtest.h"
+#include "src/ir/block_builder.h"
+#include "src/ir/datalog/value.h"
+#include "src/ir/module.h"
+
+namespace raksha::backends::policy_engine::souffle {
+
+using testing::Combine;
+using testing::TestWithParam;
+using testing::ValuesIn;
+
+struct IrAndDatalogOperationPairs {
+  const ir::Operation *ir_operation;
+  const ir::datalog::IsOperationFact *datalog_is_operation;
+};
+
+class IrOperationLoweringTest
+    : public TestWithParam<IrAndDatalogOperationPairs> {};
+
+// Check that the IrOperation is lowered to the expected `IsOperationFact`.
+// Rather than introducing a `operator==` for this, we just rely on them
+// producing the same `ToDatalogString` result.
+TEST_P(IrOperationLoweringTest, IrOperationLoweringTest) {
+  const auto [ir_operation, datalog_is_operation] = GetParam();
+
+  ir::SsaNames ssa_names;
+  DatalogLoweringVisitor datalog_lowering_visitor(ssa_names);
+  ir_operation->Accept(datalog_lowering_visitor);
+  const RakshaDatalogFacts &facts = datalog_lowering_visitor.datalog_facts();
+  EXPECT_EQ(facts.ToDatalogString(), datalog_is_operation->ToDatalogString());
+}
+
+static const ir::Operator kLiteralOperator("sql.ReadLiteral");
```

Comment:
Didn't we decide that we don't need `sql.ReadLiteral`? We can simply make it an access path?

Reply:
We do need `ReadLiteral`. What we decided that we do not need (at least for now) is an `Operation` describing source columns.